[CupertinoAlertDialog] Rewrite

dkwingsmt opened the pull request:
This PR rewrites `CupertinoAlertDialog` in a cleaner logic, mostly its layout.

There are two major difficulties to lay out `CupertinoAlertDialog`:
* Laying out the actions section, which switches between a horizontal mode (two buttons in a row) and a vertical mode (several buttons in a column). This PR rewrites it in a special layout widget, `_CupertinoAlertActionSection`.
* Allocating vertical space between the content section and the actions section. This reuses `_PriorityColumn`, which was created for the action sheet.

In a similar fashion to the action sheet, the management and configuration for dividers and background (overscrolls) were rewritten as well.

This PR tries to keep as much original code and behavior as possible since this PR is already very large. As a result, almost no tests are broken. Further improvement will be done in future PRs.
* The test that verifies painting overscrolls is replaced by a golden test, since the original test assumes that the background is painted in one rectangle.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

dkwingsmt committed:
First impl

dkwingsmt committed:
Add back state change

dkwingsmt committed:
Use original colors

dkwingsmt committed:
Thickness as argument

dkwingsmt committed:
Rewrite min/max height in widget

dkwingsmt committed:
Fix: action section rename; non-negative max height

dkwingsmt committed:
Remove media padding

dkwingsmt committed:
Fix divider when there's only title

dkwingsmt committed:
Handle content max height simpler

dkwingsmt committed:
Rename back

dkwingsmt committed:
Revert _buildContent maxHeight defining location

dkwingsmt committed:
Split out _CupertinoDialogRenderWidget

dkwingsmt committed:
Remove _CupertinoAlertActionSection's unnecessary properties

dkwingsmt committed:
Fix lint

dkwingsmt committed:
Revert back test

dkwingsmt committed:
Remove extra arg of _debugCheckConstraints

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Use getMinIntrinsicHeight to determine action section height

dkwingsmt committed:
Remove IntrinsicHeight of action button

dkwingsmt committed:
_PriorityColumn

dkwingsmt committed:
Fix some tests

dkwingsmt committed:
Add more tests

dkwingsmt committed:
Test

dkwingsmt committed:
Code

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into as-priority-column

dkwingsmt committed:
Better test doc

dkwingsmt committed:
Fix actions not being scrolled

dkwingsmt committed:
Apply suggestions from code review

Co-authored-by: Bruno Leroux <bruno.leroux@gmail.com>

dkwingsmt committed:
Merge branch 'master' into as-priority-column

dkwingsmt committed:
Merge branch 'master' into as-priority-column

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into as-priority-column

dkwingsmt committed:
Better test name

dkwingsmt committed:
Merge branch 'as-priority-column' of github.com:dkwingsmt/flutter into as-priority-column

dkwingsmt committed:
Merge branch 'as-priority-column' into ad-rewrite

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Revert some changes

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2250,26 +1937,21 @@ class CupertinoDialogAction extends StatelessWidget {
-                child: content,
-              ),
-            ),
+    return FittedBox(
```
I don't know what the `IntrinsicHeight` and `SizedBox` were for, but removing them does not break anything even without the rest of the change in this PR.

dkwingsmt committed:
Better doc

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt commented:
cc @chunhtai If you're interested in reviewing, since it's very much related to action sheets. But no pressure!

dkwingsmt committed:
Use max intrinsics width

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Fix doc

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

LongCatIsLooong reviewed (commented):
LGTM with nits. Is it possible to further simplify this by using `OverflowBar`?

dkwingsmt commented:
> Is it possible to further simplify this by using OverflowBar?

I've considered `OverflowBar` during the development and I think the two widgets actually have very different layout logic. For example, `OverflowBar` only needs lay out its children with a fixed maxWidth of containerWidth before placing them, while the alert dialog has to first query the children's intrinsic widths before deciding what constraint the children must be laid out with.

(Now that I think about it again maybe there is a way to avoid the 2nd render if we move the logic for FittedBox from the buttons to the layout widget. But even with this approach, the layout widget is quite different from the `OverflowBar`. I can give it a try in the future.)

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1299,19 +1310,24 @@ class _ActionSheetButtonBackgroundState extends State<_ActionSheetButtonBackgrou
-      decoration: BoxDecoration(
-        color: hidden ? backgroundColor : dividerColor,
+    final Color backgroundColor = CupertinoDynamicColor.resolve(hiddenColor, context);
+    return LimitedBox(
```
Oh yeah, I should have added a comment here. Basically, the `LimitedBox` allows this widget to be used as both a horizontal divider and a vertical divider without any widget-level configuration.
* In the horizontal layout, the layout widget gives the divider a max height (derived from the height of the buttons), therefore `LimitedBox` is a no-op.
* In the vertical layout, the layout widget doesn't give the divider any max height (and specifies e max width instead), therefore the height can be given by `LimitedBox`.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1510,490 +1528,126 @@ class _ActionSheetMainSheetState extends State<_ActionSheetMainSheet> {
-    _dividerPaint.color = newValue;
-    markNeedsPaint();
-  }
+class _CupertinoDialogRenderWidgetState extends State<_CupertinoDialogRenderWidget> {
```
The original doc is over-complicated and I've rewritten them as the doc in `_AlertDialogActionsLayout`.
* The original doc mixes two concepts: which mode the buttons are laid out in, and how the full height is assigned between the actions section and the content section. The two algorithms are now handled by two classes with this PR (see `_PriorityColumn`, which describes the height assignment algorithm in a more precise way).
* Also, `_CupertinoDialogRenderWidget` is now a higher-level widget that lays out the two sections and the implementation is delegated to its children widgets. This is why the doc for implementation is also moved downward.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1510,490 +1528,126 @@ class _ActionSheetMainSheetState extends State<_ActionSheetMainSheet> {
+  // necessary because the action section's background is rendered by the
+  // buttons, so that a button's background can be _replaced_ by a different
+  // color when the button is pressed.
+  Widget _buildOverscroll(Color backgroundColor) {
```
Absolutely! I planned to do this in a future PR to avoid making this PR even harder to understand, but now that you've noticed it I can refactor it right now.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2395,121 +2060,56 @@ class CupertinoDialogAction extends StatelessWidget {
-       super(children: actionButtons);
+// The layout operates in two modes:
+//
+// 1. Horizontal Mode: If there are exactly two buttons and they fit in a single
```
> I see the logic is kinda inherited from _RenderCupertinoDialogActions

Nah, the behavior is observed on iOS simulator, and the logic is what I consider the simplest way to describe the behavior.

> build the widget tree based on the number of buttons (either Column or OverflowBar)

I would if the two-button layout logic can be described as purely widgets (including `OverflowBar`). However, since now I have yet to find such a way, and so that I had to write a custom layout widget, I want to at least reuse as much existing code as possible. Now that the vertical layout is simply a `Column` for two buttons, `_AlertDialogActionsLayout` will inherit `Column` anyway, why not make `_AlertDialogActionsLayout` handle the other cases, which are also just a `Column`?

LongCatIsLooong commented:
> I've considered `OverflowBar` during the development and I think the two widgets actually have very different layout logic. For example, `OverflowBar` only needs lay out its children with a fixed maxWidth of containerWidth before placing them, while the alert dialog has to first query the children's intrinsic widths before deciding what constraint the children must be laid out with.

Ah I see, so the problem is that in horizontal mode both buttons have to be the exact same size?

dkwingsmt commented:
Yeah. If there are two buttons and they fit into 30% and 70% of the container width respectively, the dialog would use the vertical mode, while `OverflowBar` will use the horizontal mode.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2520,462 +2120,151 @@ class _RenderCupertinoDialogActions extends RenderBox
+      // If both children fit into a half-row slot, use the horizontal layout.
+      // Max intrinsic widths are used here, which, according to
+      // [TextPainter.maxIntrinsicWidth], allows text to be displayed at their
+      // full height.
```
I guess the better way is "full font size", which I've fixed to.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2094,16 +1748,22 @@ class _CupertinoAlertContentSection extends StatelessWidget {
+  //
+  // This list must not include the dividers between the buttons. If the list
+  // is empty, then this widget returns an empty box.
+  final List<Widget> actions;
```
Good idea!

dkwingsmt committed:
Some comments addressed

dkwingsmt commented:
I've tried the approach described in my previous comment...

> (Now that I think about it again maybe there is a way to avoid the 2nd render if we move the logic for FittedBox from the buttons to the layout widget. Simply put, all buttons will always be constraint with `maxWidth: fontSizeRatio * (dialogWidth - (2 * padding))`. For now this approach still feels quite different from the `OverflowBar`, but I can give it a try anyway in the future.)

And it turns out it's not feasible. This approach requires that the action widget given to the dialog widget only occupies the center part, and must not define anything that occupies the full button. For example, the action widget can't specify the mouse cursor, which is pretty counter-intuitive.

And following this convention, if we determine that the action widget should be able to define properties that occupy the full button (such as the mouse cursor or the background color), the two-pass layout is unavoidable.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -1299,19 +1310,24 @@ class _ActionSheetButtonBackgroundState extends State<_ActionSheetButtonBackgrou
+      child: Container(
+        height: _kDividerThickness,
+        decoration: BoxDecoration(
+          color: hidden ? backgroundColor : dividerColor,
```
Great catch!

dkwingsmt committed:
better divider doc

dkwingsmt committed:
Move backgorund's state to parent

dkwingsmt committed:
Refactor: combine layout class

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ad-rewrite

dkwingsmt committed:
Better doc

dkwingsmt commented:
@LongCatIsLooong I've addressed all the comments. The following are major changes:
* The overscroll background is extrated as a widget `_OverscrollBackground`.
* As a result, `_CupertinoDialogRenderWidget` became a stateless widget, and has thus been merged into `_CupertinoAlertDialogState` as method `_buildBody`.
* `_AlertDialogButtonBackground` now relies on the parent to provide `pressed` state, and becomes a stateless widget.
* Better explanation is added to `_Divider`.

dkwingsmt committed:
Move state to top

dkwingsmt committed:
Fix analyze and remove container

dkwingsmt committed:
Better _Divider

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -298,63 +306,128 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+      setState(() {
+        _pressedIndex = actionIndex;
+      });
+    } else {
```
uber nit: `else if`?

dkwingsmt replied:
I think nested `if` is more clear here, since it's basically saying "in the releasing scenario, let's check the index".

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
-      return _computeMinIntrinsicHeightSideBySide(width);
+  double computeMaxIntrinsicHeight(double width) {
+    if (!_useHorizontalLayout(width)) {
+      return super.computeMinIntrinsicHeight(width);
```
`maxIntrinsicHeight`?

dkwingsmt replied:
Great catch!

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
+    while (true) {
+      slot.layout(BoxConstraints.tight(Size(slotWidth, height)), parentUsesSize: true);
+      (slot.parentData! as FlexParentData).offset = Offset(x, 0);
+      x += slot.size.width;
```
I think in RTL + horizontal mode the first button should appear on the right? I remember RenderFlex cares about `TextDirection` when horizontal.

dkwingsmt replied:
Yeah, it's a problem raised in https://github.com/flutter/flutter/issues/99060. I plan to address this in a future PR.

dkwingsmt committed:
Fix overscroll

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
+      // Max intrinsic widths are used here, which, according to
+      // [TextPainter.maxIntrinsicWidth], allows text to be displayed at their
+      // full font size.
+      if (child.getMaxIntrinsicWidth(double.infinity) > slotWidth) {
```
nit: should this be the actual maxHeight instead of `infinity`? In case there's RenderAspectRatio or RenderWrap in the tree.

dkwingsmt replied:
If I understand correctly there isn't a maxHeight here, because in the worst case, the actions section is laid out in the vertical mode, which scrolls and allows the buttons to be as tall as they want (if they don't restrict themselves as 1 line max.)

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -2557,462 +2136,151 @@ class _RenderCupertinoDialogActions extends RenderBox
-        dividersPath.addRect(dividerRect);
+      return true;
+    }());
+    if (constraintsError != null) {
```
nit: throw after you constructed the Error in the assert?

LongCatIsLooong reviewed (approved):
LGTM

dkwingsmt committed:
Fix some comments

dkwingsmt committed:
Fix analysis
